lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -336,17 +335,31 @@ def set_aspect(self, aspect, adjustable=None, anchor=None, share=False):
             view_intervals = np.array([self.xaxis.get_view_interval(),
                                        self.yaxis.get_view_interval(),
                                        self.zaxis.get_view_interval()])
-            mean = np.mean(view_intervals, axis=1)
             ptp = np.ptp(view_intervals, axis=1)
-            delta = max(ptp[ax_indices])
-            scale = self._box_aspect[ptp == delta][0]
-            deltas = delta * self._box_aspect / scale
-
-            for i, set_lim in enumerate((self.set_xlim3d,
-                                         self.set_ylim3d,
-                                         self.set_zlim3d)):
-                if i in ax_indices:
-                    set_lim(mean[i] - deltas[i]/2., mean[i] + deltas[i]/2.)
+            if adjustable == 'datalim':
+                mean = np.mean(view_intervals, axis=1)
+                delta = max(ptp[ax_indices])
+                scale = self._box_aspect[ptp == delta][0]
+                deltas = delta * self._box_aspect / scale
+
+                for i, set_lim in enumerate((self.set_xlim3d,
+                                             self.set_ylim3d,
+                                             self.set_zlim3d)):
+                    if i in ax_indices:
+                        set_lim(mean[i] - deltas[i]/2., mean[i] + deltas[i]/2.)
+            else:  # 'box'
+                # Change the box aspect such that the ratio of the length of
+                # the unmodified axis to the length of the diagonal
+                # perpendicular to it remains unchanged.
+                box_aspect = np.array(self._box_aspect)
+                box_aspect[ax_indices] = ptp[ax_indices]
+                remaining_ax_indices = {0, 1, 2}.difference(ax_indices)
+                if remaining_ax_indices:
+                    remaining_ax_index = remaining_ax_indices.pop()
+                    old_diag = np.linalg.norm(box_aspect[ax_indices])
+                    new_diag = np.linalg.norm(ptp[ax_indices])
```

Comment:
I think these two lines should be these instead, right? Since you've already overwritten `box_aspect[ax_indices]`. The resulting plots look more how I would expect too.
```
old_diag = np.linalg.norm(self._box_aspect[ax_indices])
new_diag = np.linalg.norm(box_aspect[ax_indices])
```

Reply:
I was wondering where the calculations were going wrong (since I remember the first commit seemed fine). This was it!